### PR TITLE
Hash function included in NexusSimulator.py

### PR DESCRIPTION
**Pull request by zaclee1**
Network constraints and wells completions are the attributes used to determine uniqueness of a NexusSimulator object.

The hash function is updated to convert the attributes from a dictionary/list into tuple, and then hash it.

### CODE REVIEW

File: ResSimpy/Nexus/NexusSimulator.py [hunk 1]
```diff
@@ -143,6 +143,64 @@ def __repr__(self) -> str:
...
+    def _attr_info_to_tuple(sim_attr: Union[dict, list]) -> tuple[tuple[tuple[str, Any], ...], ...]:
+        """Convert the network constraints/wells completions attribute to a tuple of tuples so that it is hashable.
+
+        Args:
+            sim_attr (Union[dict, list]): dict if network constraints attribute, list if wells completions attribute
+
+        Returns: Tuple[Tuple[Tuple[str, Union[str, float, bool]], ...], ...]: tuple of tuples
+        """
+        if isinstance(sim_attr, dict):
+            lst_of_tuples = [tuple(nexus_constraint.to_dict(add_units=False, include_nones=False).items())
```
Comment (cflynn3) [inline ResSimpy/Nexus/NexusSimulator.py:156]: This is difficult to follow, and should be split into multiple statements, the for ... if .... for statements in particular look off. Also, the code will throw an error if `sim_attr.get(wells)` returns None, so you should check for that before attempting to loop through the result of that statement.

File: ResSimpy/Nexus/NexusSimulator.py [hunk 2]
```diff
@@ -143,6 +143,69 @@ def __repr__(self) -> str:
         printable_str += self.model_files.__repr__()
         return printable_str
 
+    @staticmethod
+    def _attr_info_to_tuple(sim_attr: Union[dict, list]) -> tuple[tuple[tuple[str, Any], ...], ...]:
```
Comment (jonnyh404) [inline ResSimpy/Nexus/NexusSimulator.py:147]: We have a return type here of tuple[str, Any] in the tuple. If that Any is a mutable data type I think the hashing won't work? Can you make sure it is tuple[str, bool, float] at that level?

File: ResSimpy/Nexus/NexusSimulator.py [hunk 3]
```diff
@@ -143,6 +143,69 @@ def __repr__(self) -> str:
...
+
+        Returns:
+            Union[str, tuple]: tuple of the network constraints and wells completions attributes or a string
+            if both the attributes return empty
+        """
+        network_attr = self.network.constraints.get_all()
+        wells_attr = self.wells.get_all()
+
+        if not network_attr and not wells_attr:
+            return "Network constraints and wells completions are empty. No hash value generated."
```
Comment (jonnyh404) [inline ResSimpy/Nexus/NexusSimulator.py:189]: We could maybe have this as returning the empty tuple and then hash the empty tuple?

File: tests/Nexus/nexus_simulator/test_nexus_simulator.py [hunk 4]
```diff
@@ -1911,3 +1911,138 @@ def test_get_fluid_type(surface_file_contents, expected_result):
...
+    mocker.patch('ResSimpy.Nexus.NexusSimulator', mock_simulator1)
+    mock_network_wells_tuple1 = mocker.MagicMock(return_value=return_value1)
+    sim1 = NexusSimulator('test/path/file1.fcs')
+    mocker.patch.object(sim1, 'network_wells_tuple', mock_network_wells_tuple1)
+
+    mock_simulator2 = mocker.MagicMock()
+    mocker.patch('ResSimpy.Nexus.NexusSimulator', mock_simulator2)
+    mock_network_wells_tuple2 = mocker.MagicMock(return_value=return_value2)
+    sim2 = NexusSimulator('test/path/file1.fcs')
+    mocker.patch.object(sim2, 'network_wells_tuple', mock_network_wells_tuple2)
```
Comment (jonnyh404) [inline tests/Nexus/nexus_simulator/test_nexus_simulator.py:2029]: This is quite aggressive mocking and don't really represent what a NexusSimulator would have. This test is really just testing if an arbitrary function with the same return value would be equal, which we would expect. You should do this with a true NexusSimulator with NexusContraints defined on it.

File: tests/Nexus/nexus_simulator/test_nexus_simulator.py [hunk 5]
```diff
@@ -1911,3 +1911,138 @@ def test_get_fluid_type(surface_file_contents, expected_result):
...
+    mock_simulator = mocker.MagicMock()
+    mocker.patch('ResSimpy.Nexus.NexusSimulator', mock_simulator)
+    mock_simulator.network_wells_tuple.return_value = (("tuple1", "tuple3"), ("tuple2", "tuple4"))
+    expected_result = hash((("tuple1", "tuple3"), ("tuple2", "tuple4")))
+
+    # Act
+    result = NexusSimulator.__hash__(mock_simulator)
+
+    # Assert
+    assert result == expected_result
```
Comment (jonnyh404) [inline tests/Nexus/nexus_simulator/test_nexus_simulator.py:1995]: Can you write a unit test that returns the hash of an unmocked NexusSimulator?

File: tests/Nexus/nexus_simulator/test_nexus_simulator.py [hunk 6]
```diff
@@ -1911,3 +1956,174 @@ def test_get_fluid_type(surface_file_contents, expected_result):
...
+    fake_simulator._wells = completions
+
+    # Act
+    result_hash = hash(fake_simulator)
+
+    # Assert
+    assert result_hash == hash(((), ()))
+
+
+def test_eq(mocker):
```
Comment (cflynn3) [inline tests/Nexus/nexus_simulator/test_nexus_simulator.py:2067]: This test is good, but I think you could do with test cases where a) both simulators are the same and b) more 'slight differences' that return false.

File: ResSimpy/Nexus/NexusSimulator.py [hunk 7]
```diff
@@ -143,6 +143,54 @@ def __repr__(self) -> str:
...
+        return network_tuple, wells_tuple
+
+    def __hash__(self) -> int:
+        hash_attr_tuple = self.network_wells_tuple()
+        return hash(hash_attr_tuple)
+
+    def __eq__(self, other) -> bool:
+        if isinstance(other, NexusSimulator):
+            return self.network_wells_tuple() == other.network_wells_tuple()
+        return NotImplemented
```
Comment (cflynn3) [inline ResSimpy/Nexus/NexusSimulator.py:192]: You should add an error message to this raised exception.

**Review by cflynn3 — APPROVED** [review]
These changes look good, just a couple of minor suggestions.

File: ResSimpy/Nexus/NexusSimulator.py [hunk 8]
```diff
@@ -166,45 +166,53 @@ def _attr_info_to_tuple(sim_attr: Union[dict, list]) -> tuple[tuple[tuple[str, A
...
-        if isinstance(hash_attr_tuple, tuple):
-            return hash(hash_attr_tuple)
-        # mypy suggests that hash return value can only be int,
-        # so it will return a hash value of 0 if both attributes are empty
-        else:
-            return 0
+        return hash(hash_attr_tuple)
 
-    def __eq__(self, other) -> bool:
+    def wells_and_network_equal(self, other) -> bool:
```
Comment (jonnyh404) [inline ResSimpy/Nexus/NexusSimulator.py:193]: Good name changes and refactor to make this clearer for the hash's actual purpose.

File: tests/Nexus/nexus_simulator/test_nexus_simulator.py [hunk 9]
```diff
@@ -1936,113 +1986,173 @@ def test_attr_info_to_tuple(mocker):
...
-    network_attr = NexusSimulator._attr_info_to_tuple(nexus_constraint_dict)
-    wells_attr = NexusSimulator._attr_info_to_tuple(nexus_well_lst)
+    network_attr = fake_simulator._attr_info_to_tuple(nexus_constraint_dict)
+    wells_attr = fake_simulator._attr_info_to_tuple(nexus_well_lst)
+    result_hash = fake_simulator.hash_network_wells()
 
     # Assert
     assert network_attr == expected_network
     assert wells_attr == expected_wells
+    assert result_hash == hash((network_attr, wells_attr))
```
Comment (jonnyh404) [inline tests/Nexus/nexus_simulator/test_nexus_simulator.py:2014]: Good test!

File: tests/Nexus/nexus_simulator/test_nexus_simulator.py [hunk 10]
```diff
@@ -1936,113 +1986,173 @@ def test_attr_info_to_tuple(mocker):
...
-    result = sim1.__eq__(sim2)
+    same_result = fake_simulator1.wells_and_network_equal(fake_simulator1)
+    diff_result = fake_simulator1.wells_and_network_equal(fake_simulator2)
 
     # Assert
-    assert result == expected_result
+    assert same_result is True
+    assert diff_result is False
+    with pytest.raises(TypeError):
+        fake_simulator1.wells_and_network_equal("Not a NexusSimulator object")
```
Comment (jonnyh404) [inline tests/Nexus/nexus_simulator/test_nexus_simulator.py:2118]: Good test!

**Review by jonnyh404 — APPROVED** [review]
Looks good to me. Good improvements to test quality + coverage

**Review by cflynn3 — APPROVED** [review]
These updated changes look good.

**Pull request merged**